### ISSUE TITLE
panic: bug: ClusterConfig called on closed or nonexistent connection 

### ISSUE THREAD

**Issue by eyeson** [syncthing/syncthing#4170]
I have had two different Syncthing instances crash on Windows 10. One is running v0.14.28, one v.0.14.29-rc1.
Both are using SyncTrayzor (V1.1.16).

I restarted one, only to have it crash ~5 mins later. I captured the Synctrazor log from that one and have attached it.

I will endeavour to provide any other information required.

Forum post for reference: [https://forum.syncthing.net/t/syncthing-crashing/9845](url)

[SyncTrayzor.txt](https://github.com/syncthing/syncthing/files/1021902/SyncTrayzor.txt)

**Comment by calmh** [comment]
For clarity:

```
[5NSF2] 09:17:06 INFO: Failed to exchange Hello messages with K2ECF46 (95.97.92.221:22067): read tcp 192.168.1.50:61924->95.97.92.221:22067: i/o timeout 
entWatcher: Received 0 events 
panic: bug: ClusterConfig called on closed or nonexistent connection 
goroutine 4242 [running]: 
github.com/syncthing/syncthing/lib/model.(*Model).ClusterConfig(0xc042058280, 0xdbf211def3228856, 0xc92a31f0bf8e42e1, 0xd87e29914adce15e, 0x87221d62cccbc3d9, 0xc042631f00, 0x1, 0x1) 
	c:/jenkins/workspace/syncthing-release-windows/src/github.com/syncthing/syncthing/lib/model/model.go:789 +0x21c3 
github.com/syncthing/syncthing/lib/protocol.(*nativeModel).ClusterConfig(0xc042b91fe0, 0xdbf211def3228856, 0xc92a31f0bf8e42e1, 0xd87e29914adce15e, 0x87221d62cccbc3d9, 0xc042631f00, 0x1, 0x1) 
	<autogenerated>:74 +0x95 
github.com/syncthing/syncthing/lib/protocol.(*rawConnection).readerLoop(0xc0457315c0, 0x0, 0x0) 
	c:/jenkins/workspace/syncthing-release-windows/src/github.com/syncthing/syncthing/lib/protocol/protocol.go:306 +0x11de 
created by github.com/syncthing/syncthing/lib/protocol.(*rawConnection).Start 
	c:/jenkins/workspace/syncthing-release-windows/src/github.com/syncthing/syncthing/lib/protocol/protocol.go:171 +0x4a 
goroutine 1 [chan receive, 70 minutes]: 
main.syncthingMain(0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0xc042057a40, 0x34, 0x0, 0x0, ...) 
	c:/jenkins/workspace/syncthing-release-windows/src/github.com/syncthing/syncthing/cmd/syncthing/main.go:954 +0x1c5a 
main.main() 
	c:/jenkins/workspace/syncthing-release-windows/src/github.com/syncthing/syncthing/cmd/syncthing/main.go:459 +0x541 
goroutine 20 [select, 71 minutes]: 
github.com/syncthing/syncthing/vendor/github.com/xtaci/kcp-go.(*updateHeap).updateTask(0x10d0f40) 
```

**Comment by calmh** [comment]
I think the connection failure at the top is relevant. Maybe this is something simple as two slow concurrent connections (i.e., one in each direction), one succeeding first and getting added, the other failing and removing the first from the model, then the first one continuing and exploding like this.

**Comment by calmh** [comment]
Note same in #3633 which this is really a dup of, but I'm leaving it open just because

**Comment by AudriusButkevicius** [comment]
I think we should just remove the panic and ignore the message, no point in crashing?

**Comment by calmh** [comment]
We might as well crash as we have obviously cleaned out the connection that we are receiving messages for and have no clue about the current reality. We need to figure out why that happens and prevent it.

**Comment by calmh** [comment]
@eyeson I don't suppose you can reproduce this somewhat reliably, or know the conditions leading up to it?

**Comment by eyeson** [comment]
@calmh Sorry, unfortunately unable to reproduce. I had been doing nothing unusual prior to the crash - if it occurs again I will try and get more details

**Comment by NCDanielH** [comment]
I had this crash for the first time just now. It happened around the same time some "unexpected" disconnections occurred (laptops moving out of WiFi range), but these disconnections normally happen dozens of times per day and have not given us trouble in the past.

They don't look too helpful, but here are the [Synctrayzor logs](https://pastebin.com/U7eZii42) from the half minute before the crash.

**Comment by scienmind** [comment]
Happens to me as well, from time to time,
ST log: https://gist.github.com/anonymous/9ac8cb7329e8af3373efea3f79e51e8c

ST v.0.14.40

**Comment by calmh** [comment]
I suspect the connection replacement code, which gets more exercise since KCP is out. (I get initial connections on KCP, switched to tcp afterwards, for whatever reason as well)

On 12 Nov 2017, at 14:05, Andrey D <notifications@github.com<mailto:notifications@github.com>> wrote:


Happens to me as well, from time to time,
ST log: https://gist.github.com/anonymous/9ac8cb7329e8af3373efea3f79e51e8c

ST v.0.14.40

**Comment by AudriusButkevicius** [comment]
It's most likely us starting a connection, receiving a message, disconnecting, and then delivering the message on a separate routine that is still alive.

**Comment by scienmind** [comment]
Just in case more logs are useful: https://gist.github.com/anonymous/f581a4066da074d0c7ca07284f22a8d5

**Comment by xor-gate** [comment]
I'm having the same https://gist.github.com/xor-gate/a11933cfc27fa64435cd1554f785e307. Both have 0.14.40 (freebsd & osx if that is usefull).

**Comment by calmh** [comment]
If anyone has a process for controlled reproduction that would be great.

**Comment by xor-gate** [comment]
It only happened once for me (today), and never happened before (using syncthing two years or so). I'm not sure how it was produced, I will keep an eye on it.

**Comment by optix2000** [comment]
Had this happen to me today, when I was setting up syncthing for the first time. Nodes connected briefly and then disconnected, until one of them crashed and restarted.

Logs seem pretty interesting as the connection appears to get get rebuilt a huge amount of times.
Non-crashing node:
https://gist.github.com/optix2000/38d8b5dc4b94912d29bb6c4c2b10305c

Crashing node:
https://gist.github.com/optix2000/b30939397a6221858f94765c32e42e80

**Comment by andrewufrank** [comment]
I have this now a lot, several times a day, but cannot produce it at will. If you tell me what log you would like I can perhaps produce it.
The configuration is 4 central nodes for syncing (all connecte to all) and relatively slow (armhf odroid - hash performance about 50 M/s) and then faster (intel) boxes connected. to one of the central nodes have 500 GB, the connected outer nodes between 20 and 200 GB synced. 
what logs would you need?

**Comment by bendlas** [comment]
I just ran into this on my laptop https://hastebin.com/qisihisewu.log

Running `syncthing-0.14.54`

**Comment by imsodin** [comment]
I'd be interested in reports of this with >=v1.0.1, as #5440 might have an influence on this.

And it looks like this is a duplicate of #3906, but more info is in this issue, so I'd close #3906 (if it really is a dupe).

**Comment by imsodin** [comment]
Another report: https://forum.syncthing.net/t/cant-add/13173  
And from it's panic and one panic in this test (checked for confirmation), it looks like the issue identified in https://github.com/syncthing/syncthing/pull/5645 is indeed the culprit:

> [...] the rawConnection causing the panic only has a readerLoop running, writerLoop doesn’t exist anymore -> “race” on closing the connection.

As #5645 does fix this as a side-effect, I opened a separate PR to fix this issue, as it's much more relevant than the main purpose of #5645, which is now closed and will be opened again once the new one is merged: #5657

**Comment by calmh** [comment]
Reopen because revert
